AsyncDisplayKit/Private/_ASHierarchyChangeSet.m

```diff
@@ -198,9 +198,41 @@ - (void)_sortAndCoalesceChangeArrays
 
     // Ignore item inserts in reloaded(new)/inserted sections.
     [_ASHierarchyItemChange sortAndCoalesceChanges:_insertItemChanges ignoringChangesInSections:insertedOrReloaded];
+    
+
+    // reload itemsChanges need to be adjusted so that we access the correct indexPaths in the datasource
+    NSDictionary *insertedIndexPathsMap = [_ASHierarchyItemChange sectionToIndexSetMapFromChanges:_insertItemChanges ofType:_ASHierarchyChangeTypeInsert];
+    NSDictionary *deletedIndexPathsMap = [_ASHierarchyItemChange sectionToIndexSetMapFromChanges:_deleteItemChanges ofType:_ASHierarchyChangeTypeDelete];
+
+    for (_ASHierarchyItemChange *change in _reloadItemChanges) {
+      NSAssert(change.changeType == _ASHierarchyChangeTypeReload, @"It must be a reload change to be in here");
+      NSMutableArray *newIndexPaths = [NSMutableArray array];
+      
+      // Every indexPaths in the change need to update its section and/or row
+      // depending on all the deletions and insertions
+      for (NSIndexPath *indexPath in change.indexPaths) {
+        NSUInteger section = indexPath.section;
+        NSUInteger row = indexPath.row;
+        
+        section -= [_deletedSections countOfIndexesInRange:NSMakeRange(0, section)];
+        section += [_insertedSections countOfIndexesInRange:NSMakeRange(0, section)];
+        
+        NSIndexSet *indicesInsertedInSection = insertedIndexPathsMap[@(section).stringValue];
+        row += [indicesInsertedInSection countOfIndexesInRange:NSMakeRange(0, row)];
+        NSIndexSet *indicesDeletedInSection = deletedIndexPathsMap[@(section).stringValue];
+        row += [indicesDeletedInSection countOfIndexesInRange:NSMakeRange(0, row)];
```

Comment:
Hmm I'm pretty sure the logic here is incorrect at both the section and row levels.

`countOfIndexesInRange` won't handle cascaded inserts correctly. For instance, say you reload item 0, and insert item at 0. We expect the reloaded index to map to 1. But `countOfIndexesInRange:(0, 0) --> 0` so we'll end up having it map to 1.

Deletes should be subtracted before inserts are added. This is because the deleted indexes are "old indexes" whereas inserted indexes are new indexes. 

Typo: Line 221 should be `-=`

See [this comment](https://github.com/facebook/AsyncDisplayKit/pull/1207#issuecomment-182172720) for a thoroughly-tested example.